### PR TITLE
Support Add Task Activities from Abstract or Non-Sealed Class

### PR DESCRIPTION
**Pull request by aueloka**
This change extends `TaskHubWorker.AddTaskActivitiesFromInterface` by supporting adding activities from `abstract` or non-sealed classes.

### CODE REVIEW

**Review by cgillum — COMMENTED** [review]
I worry a little bit about this getting messy since methods on classes aren't always going to be properly designed to handle this kind of "remoting". It seems like an area where developers could run into various issues accidentally.

Would you be willing to consider a design where classes must opt certain methods into being used as interfaces, perhaps using an attribute? For example:

```csharp
public abstract class ClassClient
{
    [TaskMethod]
    public abstract Task<T[]> Run<T>();
}
```

**Comment by aueloka** [comment]
> I worry a little bit about this getting messy since methods on classes aren't always going to be properly designed to handle this kind of "remoting". It seems like an area where developers could run into various issues accidentally.
> 
> Would you be willing to consider a design where classes must opt certain methods into being used as interfaces, perhaps using an attribute? For example:
> 
> ```cs
> public abstract class ClassClient
> {
>     [TaskMethod]
>     public abstract Task<T[]> Run<T>();
> }
> ```

Yeah fair enough. Making it explicit with an attribute makes sense here. I'll do that.

**Comment by aueloka** [comment]
@cgillum , just curious, when you say "methods on classes aren't always going to be properly designed to handle this kind of remoting, what do you mean?

What differentiates class methods from interface methods in this context? Asides from the possibility of classes having private, protected or static methods?

File: src/DurableTask.Core/OrchestrationContext.cs [hunk 1]
```diff
@@ -88,12 +88,23 @@ public abstract class OrchestrationContext
         /// <returns></returns>
         public virtual T CreateClient<T>(bool useFullyQualifiedMethodNames) where T : class
         {
-            if (!typeof(T).IsInterface)
+            if (!typeof(T).IsInterface || !typeof(T).IsClass)
```
Comment (aueloka) [inline src/DurableTask.Core/OrchestrationContext.cs:91]:
```suggestion
            if (!typeof(T).IsInterface && !typeof(T).IsClass)
```

File: src/DurableTask.Core/OrchestrationContext.cs [hunk 2]
```diff
@@ -132,14 +143,24 @@ public abstract class OrchestrationContext
         /// <returns>Dynamic proxy that can be used to schedule the remote tasks</returns>
         public virtual T CreateRetryableClient<T>(RetryOptions retryOptions, bool useFullyQualifiedMethodNames) where T : class
         {
-            if (!typeof(T).IsInterface)
+            if (!typeof(T).IsInterface || !typeof(T).IsClass)
```
Comment (aueloka) [inline src/DurableTask.Core/OrchestrationContext.cs:146]:
```suggestion
            if (!typeof(T).IsInterface && !typeof(T).IsClass)
```

**Comment by aueloka** [comment]
> @cgillum , just curious, when you say "_methods on classes aren't always going to be properly designed to handle this kind of remoting_", what do you mean?
> 
> What differentiates class methods from interface methods in this context? Asides from the possibility of classes having private, protected or static methods?



> @cgillum , just curious, when you say "_methods on classes aren't always going to be properly designed to handle this kind of remoting_", what do you mean?
> 
> What differentiates class methods from interface methods in this context? Asides from the possibility of classes having private, protected or static methods?

@cgillum , just checking in on this if you don't mind sharing your thoughts.

**Comment by aueloka** [comment]
> LGTM!

Thanks @cgillum. Can you help me merge it? Or how can I get access to be able to merge pull requests I create?

**Pull request merged**